geom/type_polygon.go

```diff
@@ -33,50 +22,85 @@ type Polygon struct {
 // any inner rings follow. If no rings are provided, then the returned Polygon
 // is the empty Polygon. The coordinate type of the polygon is the lowest
 // common coordinate type of its rings.
+//
+// An error is returned if any of the Polygon constraints are not met (see the
+// Validate method for details). Note that the validity of each input
+// LineString is not verified.
 func NewPolygon(rings []LineString, opts ...ConstructorOption) (Polygon, error) {
-	if len(rings) == 0 {
-		return Polygon{}, nil
-	}
-
-	ctype := DimXYZM
-	for _, r := range rings {
-		ctype &= r.CoordinatesType()
+	ctype := DimXY
+	if len(rings) > 0 {
+		ctype = DimXYZM
+		for _, r := range rings {
+			ctype &= r.CoordinatesType()
+		}
 	}
 	rings = append([]LineString(nil), rings...)
 	for i := range rings {
 		rings[i] = rings[i].ForceCoordinatesType(ctype)
 	}
+	poly := Polygon{rings, ctype}
 
-	ctorOpts := newOptionSet(opts)
-	if err := validatePolygon(rings, ctorOpts); err != nil {
+	os := newOptionSet(opts)
+	if os.skipValidations {
+		return poly, nil
+	}
+	if err := poly.checkPolygonConstraints(); err != nil {
 		return Polygon{}, err
 	}
-	return Polygon{rings, ctype}, nil
+	return poly, nil
+}
+
+// Validate checks if the Polygon is valid. For it to be valid, the following
+// assertions must hold:
+//
+// 1. The rings (outer and inner) must be valid linear rings. This means that
+// they must be non-empty, simple, and closed.
+//
+// 2. Each pair of rings must only intersect at a single point.
+//
+// 3. The interior of the polygon must be connected.
+//
+// 4. The holes must be fully inside the outer ring.
+//
+// In addition to these constraints, this method also verifiers that the rings
+// are valid LineStrings (i.e. have at least 2 distinct XY values, and don't
+// contain NaN or Infinite values).
+func (p Polygon) Validate() error {
+	for i, r := range p.rings {
+		if err := r.Validate(); err != nil {
+			return wrap(err, "validating ring at index %d", i)
+		}
+	}
+	err := p.checkPolygonConstraints()
+	return wrap(err, "validating polygon constraints")
 }
 
-func validatePolygon(rings []LineString, opts ctorOptionSet) error {
-	if len(rings) == 0 || opts.skipValidations {
+func (p Polygon) checkPolygonConstraints() error {
+	if len(p.rings) == 0 {
 		return nil
 	}
 
-	for _, r := range rings {
+	for _, r := range p.rings {
+		if r.IsEmpty() {
+			return defyRingNotEmpty.err()
+		}
 		if !r.IsClosed() {
-			return validationError{"polygon ring not closed"}
+			return defyRingClosed.errAtXY(r.Coordinates().GetXY(0))
 		}
 		if !r.IsSimple() {
-			return validationError{"polygon ring not simple"}
+			return defyRingSimple.errAtXY(r.Coordinates().GetXY(0))
```

Comment:
- [ ] TODO: use the point where the polygon is actually not simple, rather than just an arbitrary point